epaint/src/shape.rs

```diff
@@ -464,3 +479,902 @@ fn dashes_from_line(
+                max_x = x;
+            }
+        });
+
+        quadratic_for_each_local_extremum(self.points[0].y, self.points[1].y, self.points[2].y, &mut |t|{
+            let y = self.sample(t).y;
+            if y < min_y {
+                min_y = y;
+            }
+            if y > max_y {
+                max_y = y;
+            }
+        });
+        
+        Rect {
+            min: Pos2 { x: min_x, y: min_y },
+            max: Pos2 { x: max_x, y: max_y },
+        }
+    }
+
+    /// Calculate the point (x,y) at t based on the quadratic bezier curve equation.
+    /// t is in [0.0,1.0]
+    /// https://en.wikipedia.org/wiki/B%C3%A9zier_curve#Quadratic_B.C3.A9zier_curves
+    /// 
+    pub fn sample(&self, t: f32) -> Pos2 {
+        crate::epaint_assert!(
+            t >= 0.0 && t <= 1.0,
+            "the sample value should be in [0.0,1.0]"
+        );
+
+        let h = 1.0 - t;
+        let a = t * t;
+        let b = 2.0 * t * h;
+        let c = h * h;
+        let result = self.points[2].to_vec2() * a
+            + self.points[1].to_vec2() * b
+            + self.points[0].to_vec2() * c;
+        result.to_pos2()
+    }
+
+    /// find a set of points that approximate the quadratic bezier curve.
+    /// the number of points is determined by the tolerance.
+    /// the points may not be evenly distributed in the range [0.0,1.0] (t value)
+    pub fn flatten(&self, tolerance:Option<f32>)->Vec<Pos2>{
+        let tolerance =tolerance.unwrap_or( (self.points[0].x-self.points[2].x).abs()*0.001);
+        let mut result = Vec::new();
+        self.for_each_flattened_with_t(tolerance, &mut |p,_t|{
+            result.push(p);
+        });
+        result
+    }
+
+    // copied from https://docs.rs/lyon_geom/latest/lyon_geom/
+    /// Compute a flattened approximation of the curve, invoking a callback at
+    /// each step.
+    ///
+    /// The callback takes the point and corresponding curve parameter at each step.
+    ///
+    /// This implements the algorithm described by Raph Levien at
+    /// <https://raphlinus.github.io/graphics/curves/2019/12/23/flatten-quadbez.html>
+    pub fn for_each_flattened_with_t<F>(&self, tolerance: f32, callback: &mut F)
+    where
+        F: FnMut(Pos2, f32),
+    {
+        let params = FlatteningParameters::from_curve(self, tolerance);
+        if params.is_point {
+            return;
+        }
+
+        let count = params.count as u32;
+        for index in 1..count {
+            let t = params.t_at_iteration(index as f32);
+            
+            callback(self.sample(t),t);
+        }
+
+        callback(self.sample(1.0),1.0);
+    }
+}
+
+impl From<QuadraticBezierShape> for Shape {
+    #[inline(always)]
+    fn from(shape: QuadraticBezierShape) -> Self {
+        Self::QuadraticBezier(shape)
+    }
+}
+
+// lyon_geom::flatten_cubic.rs
+// copied from https://docs.rs/lyon_geom/latest/lyon_geom/
+pub fn flatten_cubic_bezier_with_t<F: FnMut(Pos2,f32)>(
```

Comment:
I don't think these should be `pub` (or they should be member functions)

Reply:
agree, fixed.